### PR TITLE
Show "none" as an alignment option and use contextual text to clarify settings

### LINKED ISSUE #30782: [Columns Block] Make it clearer how to return to original width option after insertion
## What problem does this address?

Currently, if you try to insert a column into the Site Editor, you're able to set two different width options: Wide or Full.  However, you're not able to return to the initial state a column block was in. This makes for a confusing experience if you just wanted to try out Wide & Full but decided to return to the initial setup. Here's a video to explain:

https://user-images.githubusercontent.com/26996883/114473364-4db29d00-9bb1-11eb-8d4f-1f9498b8ab57.mov

Initial width:

Screen Shot 2021-04-12 at 5 06 00 PM

Wide width:

Screen Shot 2021-04-12 at 5 06 05 PM

Full width:
Screen Shot 2021-04-12 at 5 06 09 PM

Because you only have two options (wide & full), you're not able to get back to the original state unless you use the undo button or delete & readd. 

## What is your proposed solution?

Allow for an option to return to the initial width setting.

### LINKED ISSUE #34597: Show "none" as an alignment option and use contextual text to clarify settings.
Updates the design of the alignment dropdown to include "none" as an option (it's the same as unsetting). The center icon has been updated, and the previous icon for centering has been repurposed for indicating "None". There's help text for "None" and "Wide", to show the values for the content and wide area layout values.

<div align="center">Group 259</div>

For themes that do not support wide or fullwide alignments we render some contextual text to clarify the missing buttons:

<div align="center">Group 260</div>

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Resolves: https://github.com/WordPress/gutenberg/issues/34597
Fixes: https://github.com/WordPress/gutenberg/issues/30782

This PR is for now a rough POC to update the design of the alignment dropdown to include "none" as an option (it's the same as unsetting). The center icon has been updated, and the previous icon for centering has been repurposed for indicating "None". There's help text for "None" and "Wide", to show the values for the content and wide area layout values.

<div align="center">Group 259</div>

For themes that do not support wide or fullwide alignments we render some contextual text to clarify the missing buttons:

<div align="center">Group 260</div>

-- the above description is from the issue

## Testing instructions
1. In a theme that supports `layout` (like `tt1-blocks`) insert a `Heading` in a post
2. Insert a `Group` and change the `layout`'s widths (`content` and `wide`)
3. Insert another `Heading` inside this `Group`
4. Check the `alignment control options` in both `Heading` blocks.

In general I will need help with testing to find out missing functionality/info, especially from folks experienced in themes.

## Tasks / Notes
- [ ] Decide on the text to show when `wide alignments` are not supported.
- [ ] Check how to calculate properly if `wide alignments` are supported
- [ ] Can we get `wide and content` widths from somewhere else besides `layout`?
- [x] Check if the align control when is a toolbar (`ToolbarGroup`) is affected somehow or needs changes
- [x] Update the `center` icon with the new one
- [ ] Check mobile
- [ ] Check whether to leave `useAvailableAlignments` intact and incorporate a new helper hook

## Notes
1. I updated the `align` tests also made sure that work as expected in `tt1 and tt1-blocks` as the button context changes. I didn't add to run on both themes though.. Should we run these for both themes? 🤔

### CODE REVIEW

File: packages/block-editor/src/components/block-alignment-control/ui.js [hunk 1]
```diff
@@ -52,43 +67,107 @@ function BlockAlignmentUI( {
...
+											onClick={ applyOrUnset( control ) }
+											role="menuitemradio"
+											info={ alignmentsInfo[ control ] }
+										>
+											{ title }
+										</MenuItem>
+									);
+								} ) }
+							</MenuGroup>
+							{ ! wideAlignmentsSupport && <div>{ help }</div> }
```
Comment (ntsekouras) [inline packages/block-editor/src/components/block-alignment-control/ui.js]: This will also need polishing like the `tag` to render and styles.

**Comment by jasmussen** [comment]
Nice work! I see some CSS variables output here:

Screenshot 2021-09-10 at 09 36 17

We need to do the icon changes as well. The existing "block-alignment-center" icon should be repurposed as the new "none" icon (in other words, renamed in the npm library). And then we add a new center icon, this one:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" style="enable-background:new 0 0 24 24" xml:space="preserve"><path d="M7 9v6h10V9H7zM5 19.8h14v-1.5H5v1.5zM5 4.3v1.5h14V4.3H5z" style="fill:#1e1e1e"/></svg>
```

Nice work!

**Comment by jasmussen** [comment]
Perhaps to explain what's going on — I have this in my theme.json. It might pose a challenge for us 🤔

```
		"layout": {
			"contentSize": "var(--wp--custom--width--center-column)",
			"wideSize": "var(--wp--custom--width--wide)"
		},
```

What, if anything, can we do here? Is referencing CSS variables inside theme.json something we should never do?

**Comment by jasmussen** [comment]
Added a Figma label as a reminder to myself.

**Comment by gziolo** [comment]
> Perhaps to explain what's going on — I have this in my theme.json. It might pose a challenge for us 🤔

Maybe it won't be that hard to parse:
https://davidwalsh.name/css-variables-javascript

File: packages/block-editor/src/components/block-alignment-control/ui.js [hunk 2]
```diff
@@ -39,7 +53,8 @@ const BLOCK_ALIGNMENTS_CONTROLS = {
 	},
 };
 
-const DEFAULT_CONTROL = 'center';
+const DEFAULT_CONTROL = 'none';
+const help = __( 'Your theme does not support wide alignments' );
```
Comment (mtias) [inline packages/block-editor/src/components/block-alignment-control/ui.js]:
```suggestion
const help = __( 'The theme does not support wider alignments.' );
```

**Comment by ntsekouras** [comment]
With this being merged: https://github.com/WordPress/gutenberg/pull/21892, do we want to have the `none` always 'highlighted'?

**Comment by mtias** [comment]
You mean when the dropdown is open?

**Comment by ntsekouras** [comment]
>You mean when the dropdown is open?

No, when it's closed.

Now I've kept it highlighted in the `dropdown` only, but for every other value it will be highlighted(background color) in the toolbar when the dropdown is closed.

**Comment by Mamaduka** [comment]
This also fixes #30782.

**Comment by ntsekouras** [comment]
>What, if anything, can we do here? Is referencing CSS variables inside theme.json something we should never do?

I don't think that is something we shouldn't do, but would like some thoughts from @oandregal or @jorgefilipecosta

**Comment by mtias** [comment]
> Now I've kept it highlighted in the dropdown only, but for every other value it will be highlighted(background color) in the toolbar when the dropdown is closed.

Ah, yes, that makes sense. I think none of the alignments should have a toggled state when resting on the toolbar? Only on the dropdown. cc @jasmussen

**Comment by oandregal** [comment]
> What, if anything, can we do here? Is referencing CSS variables inside theme.json something we should never do?

It's fine as long as the CSS variables used are the ones we generate/control via `theme.json`. In that case, we're able to parse them back to the actual value declared in theme.json. That's something we can't do if the CSS var is something we don't understand (for example, because the theme had defined it via CSS instead).

We have tools to parse the var to the actual value in the `edit-site` package (see [getValueFromVariable](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/editor/utils.js#L229)) that we're making available in the block-editor package in https://github.com/WordPress/gutenberg/pull/34178 Jorge will know more about the status of that PR and whether it's best to wait for that to land or we should use some of that code here and refactor later.

**Comment by jorgefilipecosta** [comment]
> We have tools to parse the var to the actual value in the edit-site package (see getValueFromVariable) that we're making available in the block-editor package in #34178 Jorge will know more about the status of that PR and whether it's best to wait for that to land or we should use some of that code here and refactor later.


I guess if this PR is ready to merge we can just include a temporary solution using some of the code here and then we refactor this way we keep both PR's moving without blocking each other.

File: packages/components/src/menu-item/index.js [hunk 3]
```diff
@@ -79,7 +84,7 @@ export function MenuItem(
 				className="components-menu-item__shortcut"
 				shortcut={ shortcut }
 			/>
-			{ icon && <Icon icon={ icon } /> }
+			{ icon && iconPosition === 'right' && <Icon icon={ icon } /> }
```
Comment (ciampo) [inline packages/components/src/menu-item/index.js:71]: This way of adding an icon to the right of the `Button`'s content (as opposed to the left) feels hacky — have you considered adding this functionality directly on the `Button`'s component?

**Comment by ciampo** [comment]
Not sure if I'm missing something, but I'm trying to follow the instructions and, when I add a `Heading` inside a `Group` block, I don't see the alignment options in the toolbar: (using tt1-blocks)

Screenshot 2021-09-13 at 20 09 52

**Comment by ntsekouras** [comment]
>Not sure if I'm missing something, but I'm trying to follow the instructions and, when I add a Heading inside a Group block, I don't see the alignment options in the toolbar: (using tt1-blocks)

@ciampo you missed step 2:
>Insert a Group and change the layout's widths (content and wide)

In [this PR](https://github.com/WordPress/gutenberg/pull/34586) wide/full alignments are not shown if in `flow` layout without having these values set and `Heading` [only supports wide/full](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/heading/block.json#L29)

>I guess if this PR is ready to merge we can just include a temporary solution using some of the code here and then we refactor this way we keep both PR's moving without blocking each other.

@jorgefilipecosta can you help with that or point me to the right direction? I took a look at your PR and it's pretty complex and big and in combination with the lack of mine good experience with css vars, I might need some help 😄

**Comment by jasmussen** [comment]
>  I think none of the alignments should have a toggled state when resting on the toolbar? Only on the dropdown. 

Yes, I think of it as a dropdown that also shows the selected state in its resting state. The toggled/inverted states are only inside.

**Comment by jasmussen** [comment]
Coming back to https://github.com/WordPress/gutenberg/pull/34710#issuecomment-916702784, I think it shows an edgecase that we shouldn't optimize too much for until we know it's a common thing. So as an alternative to [the proposed solution](https://github.com/WordPress/gutenberg/pull/34710#issuecomment-916715704), one thing we could perhaps do is to not output the help-text unless the format is in _number px_. Would that work?

File: packages/components/src/menu-item/index.js (hunk 3 above)
Comment (ntsekouras) [reply packages/components/src/menu-item/index.js:71]: Same comment from here: https://github.com/WordPress/gutenberg/pull/34785#discussion_r708221717

>After discussing this a bit more with @ciampo in dm and exploring the code a bit, it seems that we have been hacking our way around for many current controls. It seems that last year [I had added](https://github.com/WordPress/gutenberg/pull/26687/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-1204d0def76e28ff29950465cc842e3fafcee022c7e97c76fa5cf5baa4c4505aR119) `iconPosition` in Button but used it with `text` [added here](https://github.com/WordPress/gutenberg/pull/26425/files).

>[Tool Selector](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/tool-selector/index.js#L77) uses a similar 'hack' by passing the `icon` as a `label` along with text.

>So these usages and others that will surface by searching need to be consolidated through a more consistent API. We will probably need to make `MenuItem` always use the `icon` as a prop and also pass the `iconPosition`. This requires a greater effort and being extra cautious not to brake existing UIs and should be addressed in a separate PR.

File: packages/block-editor/src/components/block-alignment-control/use-available-alignments.js [hunk 4]
```diff
@@ -48,5 +52,10 @@ export default function useAvailableAlignments( controls = DEFAULT_CONTROLS ) {
 			availableAlignments.includes( control )
 	);
 
-	return enabledControls;
+	return {
+		enabledControls,
+		layout,
+		// TODO check if this is the check we need to show the `doesn't support wide alignments` text.
+		wideAlignmentsSupport: layout.alignments || wideControlsEnabled,
```
Comment (ntsekouras) [inline packages/block-editor/src/components/block-alignment-control/use-available-alignments.js]: I'm not sure if this is the right calculation for supported wide alignments. Folks with theme experience could help.. maybe @carolinan @aristath , @kjellr , @youknowriad

**Comment by paaljoachim** [comment]
This issue is somewhat associated:
Change default Wide width alignment for Media & Text block to normal width.
https://github.com/WordPress/gutenberg/issues/20941

File: packages/e2e-tests/specs/editor/plugins/align-hook.test.js [hunk 5]
```diff
@@ -14,121 +14,137 @@ import {
...
-		await activatePlugin( 'gutenberg-test-align-hook' );
+/**
+ * Helper function to get the `labels` of align control. It actually evaluates the
+ * basic label of the button without the `info` text node if exists.
+ *
+ * @param {Object}  options                               Options for the util function
+ * @param {boolean} [options.getActiveButtonLabels=false] Flag for returning the active buttons labels only.
+ * @return {string[]} The matched labels.
+ */
+const getAlignmentToolbarLabels = async ( {
```
Comment (ntsekouras) [inline packages/e2e-tests/specs/editor/plugins/align-hook.test.js:33]: I found these tests a bit hard to read so I moved the helper functions outside of  the main `describe`.

**Comment by carolinan** [comment]
I have tested the PR according to the testing instructions above.
The alignments are working well both in and outside containers, and also if the container itself has an alignment (chrome, windows)

_Wishlist:_
(This should not prevent the feature from being added)

If I change the content width and wide width values in the group block layout option, I would like the values to be reflected in the max width message in the inner block.
For example, using TT1 Blocks, I set the content width to 400px and wide to 1400px, but the message shows:
Max 610px and Max 1240px.


_Update:_ I see now that if I remove the layout setting from theme.json, then the message shows the value from the parent container layout option correctly.

File: packages/block-editor/src/components/block-alignment-control/use-available-alignments.js (hunk 4 above)
Comment (carolinan) [reply packages/block-editor/src/components/block-alignment-control/use-available-alignments.js]: I'm not so familiar with the code behind the alignments, but it is similar to the check inside the alignment component. I don't see any problems when I remove the layout setting from theme.json.

**Comment by jasmussen** [comment]
I added the new icons to the Figma file. What remains on this one?

**Comment by ntsekouras** [comment]
>What remains on this one?

I guess some good code review and maybe some rethinking about:
>Check whether to leave useAvailableAlignments intact and incorporate a new helper hook

File: packages/block-editor/src/components/block-alignment-control/ui.js [hunk 6]
```diff
@@ -39,7 +55,8 @@ const BLOCK_ALIGNMENTS_CONTROLS = {
 	},
 };
 
-const DEFAULT_CONTROL = 'center';
+const DEFAULT_CONTROL = 'none';
+const help = __( 'Wider alignments are not available.' );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-alignment-control/ui.js]: I actually think this message is not helpful at all, because it can confuse the user in assuming that wide alignments are not available anywhere while they are container-dependent.

File: packages/block-editor/src/components/block-alignment-control/ui.js [hunk 7]
```diff
@@ -53,42 +70,106 @@ function BlockAlignmentUI( {
...
+	const hasEnabledControls = !! enabledControls.length;
+
+	if ( ! hasEnabledControls ) {
 		return null;
 	}
+	const wideAlignmentsSupport = enabledControls.some( ( { name } ) =>
+		WIDE_CONTROLS.includes( name )
+	);
+	// Always add the `none` option if not exists.
+	if ( ! enabledControls.some( ( { name } ) => name === 'none' ) ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-alignment-control/ui.js]: Do you think it's possible to remove this and move the responsibility to the layout to always include "none". In other words, since we can't really remove "none" from the return value of `getAlignments` of layouts, we should just consider it a regular alignment with `undefined` as value.

File: packages/block-editor/src/components/block-alignment-control/use-available-alignments.js [hunk 8]
```diff
@@ -30,8 +30,9 @@ export default function useAvailableAlignments( controls = DEFAULT_CONTROLS ) {
 	const layoutAlignments = layoutType.getAlignments( layout );
 
 	if ( themeSupportsLayout ) {
-		return layoutAlignments.filter( ( control ) =>
-			controls.includes( control )
+		return layoutAlignments.filter(
+			( { name: alignmentName } ) =>
+				alignmentName === 'none' || controls.includes( alignmentName )
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-alignment-control/use-available-alignments.js]: Instead of filtering `none` here, can't we instead add it to `DEFAULT_CONTROLS` and all the places where it's missing?

File: packages/block-editor/src/components/block-alignment-control/ui.js (hunk 6 above)
Comment (ntsekouras) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: The `copy` of this is not finalized: https://github.com/WordPress/gutenberg/issues/34597#issuecomment-913925613 and IMO we could definitely could omit that here and add it in a follow up..

File: packages/block-editor/src/layouts/flow.js [hunk 9]
```diff
@@ -158,17 +158,46 @@ export default {
...
-		if ( layout.wideSize ) {
-			alignments.unshift( 'wide' );
+		// Add `none` alignment with info text.
+		if ( contentSize && ! contentSize?.startsWith( 'var' ) ) {
+			alignments.unshift( {
+				name: 'none',
+				// translators: %s: container size (i.e. 600px etc)
+				info: sprintf( __( 'Max %s wide' ), contentSize ),
+			} );
 		}
```
Comment (youknowriad) [inline packages/block-editor/src/layouts/flow.js]: Is this missing a `else` to add the `none` alignment but without `info`?

File: packages/block-editor/src/components/block-alignment-control/ui.js (hunk 6 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: I'm in the opinion that we shouldn't add it at all but 🤷‍♂️

**Review by youknowriad — COMMENTED** [review]
This is looking good in general, I left some small comments.

File: packages/block-editor/src/layouts/flow.js (hunk 9 above)
Comment (ntsekouras) [reply packages/block-editor/src/layouts/flow.js]: In this iteration no, because it was added in `BlockAlignmentUI` if wasn't present.

File: packages/block-editor/src/components/block-alignment-control/ui.js (hunk 6 above)
Comment (mtias) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: This is for cases like within a column where wide doesn't make sense? Let's skip showing anything there. We should only show a message on container where wide/full could be supported but are not.

Comment (ntsekouras) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: @mtias is it okay to skip in this PR? I think it needs more thinking and could be added in a follow up - especially if we try to add some logic to find if a block is a container or not.

Comment (mtias) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: Sounds good

File: packages/block-editor/src/components/block-alignment-control/use-available-alignments.js (hunk 8 above)
Comment (ntsekouras) [reply packages/block-editor/src/components/block-alignment-control/use-available-alignments.js]: Addressed in https://github.com/WordPress/gutenberg/pull/34710/commits/292ead1e3d9ce3646056c34a42ea1f937ff72c3d

File: packages/block-editor/src/components/block-alignment-control/ui.js (hunk 7 above)
Comment (ntsekouras) [reply packages/block-editor/src/components/block-alignment-control/ui.js]: Addressed in https://github.com/WordPress/gutenberg/pull/34710/commits/292ead1e3d9ce3646056c34a42ea1f937ff72c3d

File: packages/block-editor/src/layouts/flow.js [hunk 10]
```diff
@@ -174,32 +179,40 @@ export default {
...
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/34710#issuecomment-918000752
+ *
+ * @param {Object} layout The layout object.
+ * @return {Object} An object with contextual info per alignment.
+ */
+function getAlignmentsInfo( layout ) {
+	const { contentSize, wideSize } = layout;
+	const alignmentInfo = {};
+	if ( contentSize && ! contentSize?.startsWith( 'var' ) ) {
```
Comment (youknowriad) [inline packages/block-editor/src/layouts/flow.js]: What about `calc` and other functions? It feels instead of checking for specific function we should just check that the value is a numeric followed by an optional unit using a regex.

File: packages/block-editor/src/layouts/flow.js [hunk 11]
```diff
@@ -157,20 +157,62 @@ export default {
 		return 'vertical';
 	},
 	getAlignments( layout ) {
+		const alignmentInfo = getAlignmentsInfo( layout );
 		if ( layout.alignments !== undefined ) {
```
Comment (youknowriad) [inline packages/block-editor/src/layouts/flow.js:161]: Is this something we can deprecate maybe? (the alignment property in the flow layout, I think I added this initially before having `content` and `wide` sizes. It might not be useful anymore but I'm not sure.

File: packages/block-editor/src/layouts/flow.js (hunk 10 above)
Comment (ntsekouras) [reply packages/block-editor/src/layouts/flow.js]: You're right! I'll change it.

File: packages/block-editor/src/layouts/flow.js (hunk 11 above)
Comment (ntsekouras) [reply packages/block-editor/src/layouts/flow.js:161]: Currently we don't have a way to set this with a control, but if we set it in `code editor` it works. I haven't checked if provided in `theme.json`, but I guess this is where it could restrict some alignments..

If the `theme.json` use case doesn't make sense, I guess it's okay to remove. I'll have to check this better though..

Comment (youknowriad) [reply packages/block-editor/src/layouts/flow.js:161]: I think the use-cases don't make sense, it's just a historic thing now.

Comment (youknowriad) [reply packages/block-editor/src/layouts/flow.js:161]: I see that it's still used in some blocks (that should be refactored to flex layout): buttons, navigation.

Comment (ntsekouras) [reply packages/block-editor/src/layouts/flow.js:161]: We should revisit with the refactoring of these blocks. Sounds good?

Comment (youknowriad) [reply packages/block-editor/src/layouts/flow.js:161]: Yep, makes sense. Thank you.

**Review by youknowriad — APPROVED** [review]
I wonder if it's worth keeping the "toggling" behavior (meaning clicking the same alignment to unset it) but regardless, it looks good.

**Comment by paaljoachim** [comment]
Testing....

Media & Text block

https://user-images.githubusercontent.com/5323259/134705175-ed10c5f2-aba4-44b5-8f1d-e2526b7f1b8b.mp4

Works nicely!

I also tested with columns block with 3 columns inside a Group block. This also worked nicely!

Great work!

**Comment by ntsekouras** [comment]
Thank you all for your help and feedback!

**Comment by dcalhoun** [comment]
Hey @ntsekouras. 👋🏻 Unfortunately, these changes introduced a regression for the native rendering of `BlockAlignmentControl`. Would you be willing to review the fix in https://github.com/WordPress/gutenberg/pull/35191 to help ensure the approach is appropriate and the web UI continues to function as expected please?

**Comment by ntsekouras** [comment]
>Unfortunately, these changes introduced a regression for the native rendering of BlockAlignmentControl

Oh shoot... Looking at the PR right now!! Thanks @dcalhoun !
